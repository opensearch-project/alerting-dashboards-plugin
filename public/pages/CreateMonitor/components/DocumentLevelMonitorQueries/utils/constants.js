/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { OPERATORS_MAP } from '../../MonitorExpressions/expressions/utils/constants';

export const DOC_LEVEL_INPUT_FIELD = 'doc_level_input';

/**
 * A list of the operators currently supported for defining queries through the UI.
 */
export const SUPPORTED_DOC_LEVEL_QUERY_OPERATORS = [
  OPERATORS_MAP.IS.value,
  OPERATORS_MAP.IS_NOT.value,
  OPERATORS_MAP.IS_GREATER.value,
  OPERATORS_MAP.IS_GREATER_EQUAL.value,
  OPERATORS_MAP.IS_LESS.value,
  OPERATORS_MAP.IS_LESS_EQUAL.value,
];

/**
 * Initial implementation of document level monitors used these feature-specific operator values.
 * Refactored document level monitors assets to reuse the OPERATORS_MAP to reduce duplicate code
 * as new query types become supported.
 */
export const LEGACY_QUERY_OPERATORS = {
  IS: { text: 'is', value: '==' },
  IS_NOT: { text: 'is not', value: '!=' },
};

/**
 * These patterns delineate the field being queries from the query value
 * when using query string query syntax.
 */
export const QUERY_STRING_QUERY_OPERATORS = {
  [OPERATORS_MAP.IS_GREATER.value]: ':>',
  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: ':>=',
  [OPERATORS_MAP.IS_LESS.value]: ':<',
  [OPERATORS_MAP.IS_LESS_EQUAL.value]: ':<=',
};

/**
 * Similar to OPERATORS_QUERY_MAP, this const contains the query patterns
 * for the SUPPORTED_DOC_LEVEL_QUERY_OPERATORS.
 */
export const DOC_LEVEL_QUERY_MAP = {
  [LEGACY_QUERY_OPERATORS.IS.value]: {
    query: ({ field, query }) => `${field}:\"${query}\"`,
  },
  [LEGACY_QUERY_OPERATORS.IS_NOT.value]: {
    query: ({ field, query }) => `NOT (${field}:\"${query}\")`,
  },
  [OPERATORS_MAP.IS.value]: {
    query: ({ field, query }) => `${field}:\"${query}\"`,
  },
  [OPERATORS_MAP.IS_NOT.value]: {
    query: ({ field, query }) => `NOT (${field}:\"${query}\")`,
  },
  [OPERATORS_MAP.IS_GREATER.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_GREATER.value]}${query}`,
  },
  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_GREATER_EQUAL.value]}${query}`,
  },
  [OPERATORS_MAP.IS_LESS.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_LESS.value]}${query}`,
  },
  [OPERATORS_MAP.IS_LESS_EQUAL.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_LESS_EQUAL.value]}${query}`,
  },
};
