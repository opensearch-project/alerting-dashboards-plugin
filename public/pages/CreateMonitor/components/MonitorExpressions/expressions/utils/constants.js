/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const POPOVER_STYLE = { zIndex: '200' };
export const EXPRESSION_STYLE = { padding: '20px', whiteSpace: 'nowrap' };
export const Expressions = {
  THRESHOLD: 'THRESHOLD',
  WHEN: 'WHEN',
  OF_FIELD: 'OF_FIELD',
  OVER: 'OVER',
  FOR_THE_LAST: 'FOR_THE_LAST',
  WHERE: 'WHERE',
  METRICS: 'METRICS',
  GROUP_BY: 'GROUP_BY',
};
export const NUMBER_TYPES = [
  'long',
  'integer',
  'short',
  'byte',
  'double',
  'float',
  'half_float',
  'scaled_float',
];
export const UNITS_OF_TIME = [
  { value: 'm', text: 'minute(s)' },
  { value: 'h', text: 'hour(s)' },
  { value: 'd', text: 'day(s)' },
];

export const WHERE_BOOLEAN_FILTERS = [
  { text: 'Select value', value: '' },
  { text: 'True', value: true },
  { text: 'False', value: false },
];

export const OPERATORS_MAP = {
  IS: {
    text: 'is',
    value: 'is',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NOT: {
    text: 'is not',
    value: 'is_not',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NULL: {
    text: 'is null',
    value: 'is_null',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NOT_NULL: {
    text: 'is not null',
    value: 'is_not_null',
    dataTypes: ['number', 'text', 'keyword'],
  },
  IS_GREATER: {
    text: 'is greater than',
    value: 'is_greater',
    dataTypes: ['number'],
  },
  IS_GREATER_EQUAL: {
    text: 'is greater than equal',
    value: 'is_greater_equal',
    dataTypes: ['number'],
  },
  IS_LESS: {
    text: 'is less than',
    value: 'is_less',
    dataTypes: ['number'],
  },
  IS_LESS_EQUAL: {
    text: 'is less than equal',
    value: 'is_less_equal',
    dataTypes: ['number'],
  },
  STARTS_WITH: {
    text: 'starts with',
    value: 'starts_with',
    dataTypes: ['text', 'keyword'],
  },
  ENDS_WITH: {
    text: 'ends with',
    value: 'ends_with',
    dataTypes: ['text', 'keyword'],
  },
  CONTAINS: {
    text: 'contains',
    value: 'contains',
    dataTypes: ['text', 'keyword'],
  },
  DOES_NOT_CONTAINS: {
    text: 'does not contain',
    value: 'does_not_contains',
    dataTypes: ['text'],
  },
  IN_RANGE: {
    text: 'is in range',
    value: 'in_range',
    dataTypes: ['number'],
  },
  NOT_IN_RANGE: {
    text: 'is not in range',
    value: 'not_in_range',
    dataTypes: ['number'],
  },
};

export const OVER_TYPES = [{ value: 'all documents', text: 'all documents' }];

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const GROUP_BY_ERROR = 'Must specify at least 1 group by expression.';
export const QUERY_TYPE_GROUP_BY_ERROR = 'Can have a maximum of 1 group by selections.';

export const QUERY_TYPE_METRIC_ERROR = 'Can have a maximum of 1 metric selections.';

export const MAX_NUM_QUERY_LEVEL_GROUP_BYS = 1;
export const MAX_NUM_BUCKET_LEVEL_GROUP_BYS = 2;
export const MAX_NUM_WHERE_EXPRESSION = 1;
export const WHERE_FILTER_ALLOWED_TYPES = ['number', 'text', 'keyword', 'boolean'];
