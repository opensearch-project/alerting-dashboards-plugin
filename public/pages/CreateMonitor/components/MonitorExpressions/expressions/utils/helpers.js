/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export function selectOptionValueToText(optionValue, options) {
  return options.find((opt) => opt.value === optionValue).text;
}

export function getOfExpressionAllowedTypes(values) {
  const types = ['number'];
  if (['min', 'max'].includes(values.aggregationType)) types.push('date');
  return types;
}

export function getMetricExpressionAllowedTypes(values) {
  const types = ['number'];

  // Check if the aggregation types includes any of 'min', 'max', or 'count'.
  // If so, add 'date' to the allowed type for metric expression field.
  if (
    values.aggregations.some((e) => {
      return ['min', 'max', 'count'].includes(e.aggregationType);
    })
  )
    types.push('date');

  // Check if the aggregations types includes 'count'.
  // If so, add 'keyword' to the allowed type for metric expression field.
  if (
    values.aggregations.some((e) => {
      return ['count'].includes(e.aggregationType);
    })
  )
    types.push('keyword');
  return types;
}

export function getGroupByExpressionAllowedTypes() {
  return ['keyword'];
}

export const validateAggregationsDuplicates = (aggregations) => {
  let duplicates = 0;
  aggregations.forEach((e1, index) => {
    aggregations.slice(index + 1).forEach((e2) => {
      if (e1.aggregationType === e2.aggregationType && e1.fieldName === e2.fieldName) {
        duplicates++;
      }
    });
  });
  return duplicates > 0;
};
