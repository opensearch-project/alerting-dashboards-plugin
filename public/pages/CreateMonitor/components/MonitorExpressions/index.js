/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import MonitorExpressions from './MonitorExpressions';

export default MonitorExpressions;
