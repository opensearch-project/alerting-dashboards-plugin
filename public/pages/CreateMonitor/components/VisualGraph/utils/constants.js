/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const SIZE_RANGE = [3, 3];
export const LINE_STYLES = { strokeWidth: 1 };
export const ANNOTATION_STYLES = { strokeWidth: 1, stroke: 'red' };
export const HINT_STYLES = {
  background: '#3a3a48',
  borderRadius: '5px',
  padding: '7px 10px',
  fontSize: '12px',
  fontWeight: 'bold',
  color: '#fff',
  textAlign: 'left',
  whiteSpace: 'nowrap',
  boxShadow: '0 2px 4px rgba(0,0,0,.5)',
};

// Don't want values to reach top of graph so increase by a buffer % so we always have a yDomain with a slightly higher max
export const Y_DOMAIN_BUFFER = 1.4; // 40%

export const X_DOMAIN_BUFFER = 0.2; // 40%

// Size of circle for each point on graph
export const DEFAULT_MARK_SIZE = 3;

export const BAR_PERCENTAGE = 0.015;

export const BAR_KEY_COUNT = 3;
