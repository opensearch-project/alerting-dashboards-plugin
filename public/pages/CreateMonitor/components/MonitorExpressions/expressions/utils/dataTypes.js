/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { NUMBER_TYPES } from './constants';
import { DATA_TYPES } from '../../../../../../utils/constants';

export function getFieldsForType(dataTypes, type) {
  if (type === DATA_TYPES.NUMBER) {
    return NUMBER_TYPES.reduce(
      (options, type) => options.concat(getFieldsForType(dataTypes, type)),
      []
    );
  }
  if (dataTypes[type]) {
    return [...dataTypes[type].values()];
  }
  return [];
}

export const getIndexFields = (dataTypes, allowedTypes) =>
  allowedTypes.map((type) => ({
    label: type,
    options: getFieldsForType(dataTypes, type).map((field) => ({ label: field, type })),
  }));

export const getFilteredIndexFields = (dataTypes, allowedTypes, fieldsToInclude) =>
  allowedTypes.map((type) => ({
    label: type,
    options: getFieldsForType(dataTypes, type)
      .filter((field) => fieldsToInclude.includes(field))
      .map((field) => ({ label: field, type })),
  }));

/**
 * Searches the output from 'getIndexFields' for the provided field to determine the field's type.
 * @param queryField - a string representing the field being queried.
 * @param indexFields - the array output from 'getIndexFields'.
 * @return {*|string} - a string representing the data type of the queryField.
 */
export function getTypeForField(queryField = '', indexFields = []) {
  if (_.isEmpty(queryField) || _.isEmpty(indexFields)) return DATA_TYPES.TEXT;
  const match = indexFields.find(
    (entry) => entry.options?.find((dataField) => dataField.label === queryField) !== undefined
  );
  return match?.label || DATA_TYPES.TEXT;
}
