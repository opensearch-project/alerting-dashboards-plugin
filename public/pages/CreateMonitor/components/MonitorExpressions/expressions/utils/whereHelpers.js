/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { OPERATORS_MAP, Expressions } from './constants';
import { TRIGGER_COMPARISON_OPERATORS } from '../../../../../CreateTrigger/containers/DefineBucketLevelTrigger/DefineBucketLevelTrigger';
import { DATA_TYPES } from '../../../../../../utils/constants';
import { FORMIK_INITIAL_WHERE_EXPRESSION_VALUES } from '../../../../containers/CreateMonitor/utils/constants';

export const DEFAULT_WHERE_EXPRESSION_TEXT = 'All fields are included';

export const getOperators = (
  fieldType = DATA_TYPES.TEXT,
  supportedOperators = Object.values(OPERATORS_MAP)
) =>
  supportedOperators.reduce(
    (acc, currentOperator) =>
      currentOperator.dataTypes.includes(fieldType)
        ? [...acc, { text: currentOperator.text, value: currentOperator.value }]
        : acc,
    []
  );

export const isRangeOperator = (selectedOperator) =>
  [OPERATORS_MAP.IN_RANGE.value, OPERATORS_MAP.NOT_IN_RANGE.value].includes(selectedOperator);
export const isNullOperator = (selectedOperator) =>
  [OPERATORS_MAP.IS_NULL.value, OPERATORS_MAP.IS_NOT_NULL.value].includes(selectedOperator);

export const displayText = (whereValues) => {
  const comparisonOperators = _.concat(Object.values(OPERATORS_MAP), TRIGGER_COMPARISON_OPERATORS);

  const whereFieldName = _.get(whereValues, 'fieldName[0].label', undefined);
  if (!whereFieldName) return DEFAULT_WHERE_EXPRESSION_TEXT;

  const selectedOperator = _.get(whereValues, 'operator', OPERATORS_MAP.IS);
  const operatorObj =
    comparisonOperators.find((operator) => operator.value === selectedOperator) || {};
  const initialText = `${whereFieldName} ${operatorObj.text || ''}`;

  if (isRangeOperator(selectedOperator)) {
    const startRange = _.get(whereValues, 'fieldRangeStart');
    const endRange = _.get(whereValues, 'fieldRangeEnd');
    return `${initialText} from ${startRange} to ${endRange}`;
  } else if (isNullOperator(selectedOperator)) {
    return `${initialText}`;
  } else {
    const value = _.get(whereValues, 'fieldValue', '');
    return `${initialText} ${value}`;
  }
};

export const validateRange = (value = '', filter = FORMIK_INITIAL_WHERE_EXPRESSION_VALUES) => {
  if (_.isEmpty(value?.toString())) return 'Required.';
  if (filter.fieldRangeEnd < value || filter.fieldRangeStart === filter.fieldRangeEnd)
    return 'Start should be less than end range.';
  if (value < filter.fieldRangeStart) return 'End should be greater than start range.';
  return undefined;
};

export const validateWhereFilter = (filter = FORMIK_INITIAL_WHERE_EXPRESSION_VALUES) => {
  const fieldName = _.get(
    filter,
    'fieldName[0]',
    FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.fieldName[0]
  );
  const fieldOperator = _.get(filter, 'operator', FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.operator);
  let filterIsValid = !_.isEmpty(fieldName.label);
  switch (fieldOperator) {
    case OPERATORS_MAP.IS:
    case OPERATORS_MAP.IS_NOT:
    case OPERATORS_MAP.IS_GREATER:
    case OPERATORS_MAP.IS_GREATER_EQUAL:
    case OPERATORS_MAP.IS_LESS:
    case OPERATORS_MAP.IS_LESS_EQUAL:
    case OPERATORS_MAP.STARTS_WITH:
    case OPERATORS_MAP.ENDS_WITH:
    case OPERATORS_MAP.CONTAINS:
    case OPERATORS_MAP.NOT_CONTAINS:
      // These operators store the query value in the 'fieldValue'
      // attribute of FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.
      // Validate that value.
      filterIsValid = filterIsValid && !_.isEmpty(filter.fieldValue?.toString());
      break;
    case OPERATORS_MAP.IN_RANGE:
    case OPERATORS_MAP.NOT_IN_RANGE:
      // These operators store the query values in the 'fieldRangeStart' and 'fieldRangeEnd'
      // attributes of FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.
      // Both of those values need to be validated.
      filterIsValid =
        filterIsValid &&
        !validateRange(filter.fieldRangeStart, filter) &&
        !validateRange(filter.fieldRangeEnd, filter);
      break;
    case OPERATORS_MAP.IS_NULL:
    case OPERATORS_MAP.IS_NOT_NULL:
      // These operators don't store a query value in the FORMIK_INITIAL_WHERE_EXPRESSION_VALUES.
      // No further validation needed.
      break;
    default:
      console.log('Unknown query operator detected:', fieldOperator);
      filterIsValid = false;
  }
  return filterIsValid;
};

export const validateWhereFilters = (filters = []) => {
  return filters.filter((filter) => validateWhereFilter(filter)).length === filters.length;
};

export const getIsDataFilterActive = ({ formik: { values }, openedStates, fieldPath = '' }) => {
  const whereValues = _.get(values, `${fieldPath}where`);
  const whereFieldName = _.get(whereValues, 'fieldName[0].label', undefined);
  return openedStates[Expressions.WHERE] || whereFieldName;
};
