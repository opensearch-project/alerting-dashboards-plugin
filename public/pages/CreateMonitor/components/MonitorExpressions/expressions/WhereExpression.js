/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'formik';
import {
  EuiFlexGroup,
  EuiFlexItem,
  EuiPopover,
  EuiButtonEmpty,
  EuiText,
  EuiBadge,
  EuiSpacer,
} from '@elastic/eui';
import _ from 'lodash';
import {
  Expressions,
  POPOVER_STYLE,
  EXPRESSION_STYLE,
  WHERE_BOOLEAN_FILTERS,
} from './utils/constants';
import {
  getOperators,
  displayText,
  validateRange,
  isNullOperator,
  isRangeOperator,
} from './utils/whereHelpers';
import {
  hasError,
  isInvalid,
  required,
  validateRequiredNumber,
} from '../../../../../utils/validate';
import {
  FormikComboBox,
  FormikSelect,
  FormikFieldNumber,
  FormikFieldText,
} from '../../../../../components/FormControls';
import { getFilteredIndexFields, getIndexFields } from './utils/dataTypes';
import {
  FILTERS_TOOLTIP_TEXT,
  FORMIK_INITIAL_VALUES,
  TIME_RANGE_TOOLTIP_TEXT,
} from '../../../containers/CreateMonitor/utils/constants';
import { DATA_TYPES } from '../../../../../utils/constants';
import {
  TRIGGER_COMPARISON_OPERATORS,
  TRIGGER_OPERATORS_MAP,
} from '../../../../CreateTrigger/containers/DefineBucketLevelTrigger/DefineBucketLevelTrigger';
import { FORMIK_INITIAL_TRIGGER_VALUES } from '../../../../CreateTrigger/containers/CreateTrigger/utils/constants';
import { inputLimitText } from '../../../../../utils/helpers';
import IconToolTip from '../../../../../components/IconToolTip';

const propTypes = {
  formik: PropTypes.object.isRequired,
  dataTypes: PropTypes.object.isRequired,
  onMadeChanges: PropTypes.func.isRequired,
  openedStates: PropTypes.object.isRequired,
  openExpression: PropTypes.func.isRequired,
};

const ALLOWED_TYPES = ['number', 'text', 'keyword', 'boolean'];

const MAX_NUM_WHERE_EXPRESSION = 1;

class WhereExpression extends Component {
  constructor(props) {
    super(props);
  }

  handleFieldChange = (option, field, form) => {
    this.props.onMadeChanges();
    this.resetValues();
    form.setFieldValue(field.name, option);
    // User can remove where condition
    if (option.length === 0) {
      this.resetValues();
      form.setFieldError('where', undefined);
    }
  };

  handleOperatorChange = (e, field) => {
    this.props.onMadeChanges();
    field.onChange(e);
  };

  handleChangeWrapper = (e, field) => {
    this.props.onMadeChanges();
    field.onChange(e);
  };

  handleClosePopOver = async () => {
    const {
      formik: { values },
      closeExpression,
      fieldPath = '',
    } = this.props;
    // Explicitly invoking validation, this component unmount after it closes.
    const fieldName = _.get(values, `${fieldPath}where.fieldName`, '');
    if (fieldName > 0) {
      await this.props.formik.validateForm();
    }
    closeExpression(Expressions.WHERE);
  };

  resetValues = () => {
    const { fieldPath, formik, useTriggerFieldOperators = false } = this.props;
    if (useTriggerFieldOperators) {
      _.set(formik, `values.${fieldPath}where`, FORMIK_INITIAL_TRIGGER_VALUES.where);
      formik.setValues({ ...formik.values });
    } else {
      formik.setValues({
        ...formik.values,
        where: { ...FORMIK_INITIAL_VALUES.where },
      });
    }
  };

  renderBetweenAnd = () => {
    const {
      formik: { values },
      fieldPath = '',
    } = this.props;
    return (
      <EuiFlexGroup alignItems="center">
        <EuiFlexItem>
          <FormikFieldNumber
            name={`${fieldPath}where.fieldRangeStart`}
            fieldProps={{
              validate: (value) => validateRange(value, _.get(values, `${fieldPath}where`)),
            }}
            inputProps={{ onChange: this.handleChangeWrapper, isInvalid }}
          />
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiText textAlign="center">TO</EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <FormikFieldNumber
            name={`${fieldPath}where.fieldRangeEnd`}
            fieldProps={{
              validate: (value) => validateRange(value, _.get(values, `${fieldPath}where`)),
            }}
            inputProps={{ onChange: this.handleChangeWrapper, isInvalid }}
          />
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  renderValueField = (fieldType, fieldOperator) => {
    const { fieldPath = '' } = this.props;
    if (fieldType === DATA_TYPES.NUMBER) {
      return isRangeOperator(fieldOperator) ? (
        this.renderBetweenAnd()
      ) : (
        <FormikFieldNumber
          name={`${fieldPath}where.fieldValue`}
          fieldProps={{ validate: validateRequiredNumber }}
          inputProps={{ onChange: this.handleChangeWrapper }}
          formRow
          rowProps={{ isInvalid, error: hasError }}
        />
      );
    } else if (fieldType === DATA_TYPES.BOOLEAN) {
      return (
        <FormikSelect
          name={`${fieldPath}where.fieldValue`}
          fieldProps={{ validate: required }}
          inputProps={{
            onChange: this.handleChangeWrapper,
            options: WHERE_BOOLEAN_FILTERS,
            isInvalid,
          }}
        />
      );
    } else {
      return (
        <FormikFieldText
          name={`${fieldPath}where.fieldValue`}
          fieldProps={{ validate: required }}
          inputProps={{ onChange: this.handleChangeWrapper, isInvalid }}
        />
      );
    }
  };

  render() {
    const {
      formik: { values },
      openedStates,
      openExpression,
      dataTypes,
      indexFieldFilters,
      useTriggerFieldOperators = false,
      fieldPath = '',
    } = this.props;

    const indexFields =
      indexFieldFilters !== undefined
        ? getFilteredIndexFields(dataTypes, ALLOWED_TYPES, indexFieldFilters)
        : getIndexFields(dataTypes, ALLOWED_TYPES);
    const fieldType = _.get(values, `${fieldPath}where.fieldName[0].type`, 'number');
    let fieldOperator = _.get(values, `${fieldPath}where.operator`, 'is');
    if (useTriggerFieldOperators && !_.includes(_.values(TRIGGER_OPERATORS_MAP), fieldOperator)) {
      fieldOperator = TRIGGER_OPERATORS_MAP.INCLUDE;
      _.set(values, `${fieldPath}where.operator`, fieldOperator);
    }

    const fieldOperators = useTriggerFieldOperators
      ? TRIGGER_COMPARISON_OPERATORS
      : getOperators(fieldType);

    const whereFilterHeader = useTriggerFieldOperators ? 'Keyword filter' : 'Data filter';

    const whereValues = _.get(values, `${fieldPath}where`);
    const whereFieldName = _.get(whereValues, 'fieldName[0].label', undefined);

    const showAddButtonFlag = !openedStates[Expressions.WHERE] && !whereFieldName;

    return (
      <div>
        <EuiText size="xs">
          <strong>{whereFilterHeader}</strong>
          <i> - optional </i>
          <IconToolTip content={FILTERS_TOOLTIP_TEXT} iconType="questionInCircle" />
        </EuiText>
        <EuiSpacer size={'s'} />

        {showAddButtonFlag ? (
          <div>
            <EuiText size={'xs'}>No filters defined.</EuiText>
          </div>
        ) : (
          <EuiPopover
            id={`${whereFilterHeader}-badge-popover`}
            button={
              <div style={{ paddingBottom: '5px' }}>
                <EuiBadge
                  color="hollow"
                  iconSide="right"
                  iconType="cross"
                  iconOnClick={() => this.resetValues()}
                  iconOnClickAriaLabel="Remove where filter"
                  onClick={() => {
                    openExpression(Expressions.WHERE);
                  }}
                  onClickAriaLabel="Edit where filter"
                >
                  {displayText(_.get(values, `${fieldPath}where`))}
                </EuiBadge>
              </div>
            }
            isOpen={openedStates.WHERE}
            closePopover={this.handleClosePopOver}
            panelPaddingSize="none"
            ownFocus
            withTitle
            anchorPosition="downLeft"
          >
            <div style={POPOVER_STYLE}>
              <EuiFlexGroup style={{ ...EXPRESSION_STYLE }}>
                <EuiFlexItem grow={false} style={{ width: 200 }}>
                  <FormikComboBox
                    name={`${fieldPath}where.fieldName`}
                    inputProps={{
                      placeholder: 'Select a field',
                      options: indexFields,
                      onChange: this.handleFieldChange,
                      isClearable: false,
                      singleSelection: { asPlainText: true },
                    }}
                  />
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <FormikSelect
                    name={`${fieldPath}where.operator`}
                    inputProps={{
                      onChange: this.handleOperatorChange,
                      options: fieldOperators,
                    }}
                  />
                </EuiFlexItem>
                {!isNullOperator(fieldOperator) && (
                  <EuiFlexItem>{this.renderValueField(fieldType, fieldOperator)}</EuiFlexItem>
                )}
              </EuiFlexGroup>
            </div>
          </EuiPopover>
        )}

        {showAddButtonFlag && (
          <EuiButtonEmpty
            size="xs"
            data-test-subj={`${fieldPath}where.addFilterButton`}
            onClick={() => openExpression(Expressions.WHERE)}
            style={{ paddingTop: '5px' }}
          >
            + Add filter
          </EuiButtonEmpty>
        )}

        {inputLimitText(
          showAddButtonFlag ? 0 : 1,
          MAX_NUM_WHERE_EXPRESSION,
          _.lowerCase(whereFilterHeader),
          _.lowerCase(`${whereFilterHeader}s`),
          { paddingLeft: '10px' }
        )}
      </div>
    );
  }
}

WhereExpression.propTypes = propTypes;

export default connect(WhereExpression);
