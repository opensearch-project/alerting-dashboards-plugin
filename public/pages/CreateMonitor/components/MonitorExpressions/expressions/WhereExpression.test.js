/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { Formik } from 'formik';
import { render, mount } from 'enzyme';
import { EuiPopover } from '@elastic/eui';

import { FORMIK_INITIAL_VALUES } from '../../../containers/CreateMonitor/utils/constants';
import WhereExpression from './WhereExpression';
import { FormikFieldText, FormikFieldNumber } from '../../../../../components/FormControls';
import { OPERATORS_MAP } from './utils/constants';

const dataTypes = {
  integer: new Set(['age']),
  text: new Set(['cityName']),
  keyword: new Set(['cityName.keyword']),
};
const openExpression = jest.fn();
const closeExpression = jest.fn();
const getMountWrapper = (state = false) => (
  <Formik initialValues={FORMIK_INITIAL_VALUES}>
    {(props) => (
      <WhereExpression
        formik={props}
        dataTypes={dataTypes}
        openedStates={{ WHERE: state }}
        openExpression={openExpression}
        closeExpression={closeExpression}
        onMadeChanges={jest.fn()}
      />
    )}
  </Formik>
);

describe('WhereExpression', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders', () => {
    expect(render(getMountWrapper())).toMatchSnapshot();
  });
  test('calls openExpression when clicking expression', () => {
    const wrapper = mount(getMountWrapper());
    const button = wrapper.find('[data-test-subj="where.addFilterButton"]').first();
    button.simulate('click');
    wrapper.update();
    expect(openExpression).toHaveBeenCalled();
  });

  // TODO: Skipping this test for now. OpenSearch-Dashboards bump the version of EUI it uses when upgrading from 1.3 to 2.0.
  //  The current version refactored the EuiPopover to handle `onKeyDown` events at the document level instead of the component level.
  //  That change breaks the functionality of this test which makes use of the `escape` key to close the popover.
  //  Manually tested this behavior May 9, 2022, and the popover closes as expected.
  //  https://github.com/opensearch-project/alerting-dashboards-plugin/issues/236
  test.skip('calls closeExpression when closing popover', () => {
    const wrapper = mount(getMountWrapper(true));
    wrapper.find(EuiPopover).simulate('keyDown', { keyCode: 27 });
    expect(closeExpression).toHaveBeenCalled();
  });

  test('should render text input for the text data types', async () => {
    const wrapper = mount(getMountWrapper(true));
    wrapper
      .find('[data-test-subj="comboBoxSearchInput"]')
      .hostNodes()
      .simulate('change', { target: { value: 'cityName' } })
      .simulate('keyDown', { key: 'ArrowDown' })
      .simulate('keyDown', { key: 'Enter' })
      .simulate('blur');

    wrapper.update();
    const values = wrapper.find(WhereExpression).props().formik.values;
    expect(values.where.fieldName).toEqual([{ label: 'cityName', type: 'text' }]);
    expect(values.where.operator).toEqual(OPERATORS_MAP.IS);
    expect(wrapper.find(FormikFieldText).length).toBe(1);
    expect(wrapper.find(FormikFieldNumber).length).toBe(0);
  });
});
