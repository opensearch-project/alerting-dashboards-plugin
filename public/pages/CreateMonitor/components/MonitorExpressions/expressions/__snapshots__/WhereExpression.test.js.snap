// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhereExpression renders 1`] = `
<div>
  <div
    class="euiText euiText--extraSmall"
  >
    <strong>
      Data filter
    </strong>
    <i>
       - optional 
    </i>
    <span
      class="euiToolTipAnchor"
    >
      <svg
        aria-hidden="true"
        class="euiIcon euiIcon--medium euiIcon-isLoading"
        focusable="false"
        height="16"
        role="img"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      />
    </span>
  </div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div>
    <div
      class="euiText euiText--extraSmall"
    >
      No filters defined.
    </div>
  </div>
  <button
    class="euiButtonEmpty euiButtonEmpty--primary euiButtonEmpty--xSmall"
    data-test-subj="where.addFilterButton"
    style="padding-top:5px"
    type="button"
  >
    <span
      class="euiButtonContent euiButtonEmpty__content"
    >
      <span
        class="euiButtonEmpty__text"
      >
        + Add filter
      </span>
    </span>
  </button>
  <div
    class="euiText euiText--extraSmall"
    style="padding-left:10px"
  >
    <div
      class="euiTextColor euiTextColor--subdued"
    >
      You can add up to 1  data filter.
    </div>
  </div>
</div>
`;
